compiler/rustc_builtin_macros/src/deriving/default.rs

```diff
@@ -222,9 +222,6 @@ fn validate_default_attribute(
             "this method must only be called with a variant that has a `#[default]` attribute",
         ),
         [first, rest @ ..] => {
-            // FIXME(jhpratt) Do we want to perform this check? It doesn't exist
-            // for `#[inline]`, `#[non_exhaustive]`, and presumably others.
-
             let suggestion_text =
                 if rest.len() == 1 { "try removing this" } else { "try removing these" };
 
```

Comment:
FWIW it feels unfortunate that we're emitting this error in combination with the 'multiple declared defaults' error -- can we move the multiple declared defaults error here, and then just have a delay_span_bug! in the extraction code or similar?

(Obviously, this is not directly related to the FIXME, but it feels like if we're going to not do that then we should adjust the FIXME to say "actually do that validation").

Reply:
Yeah, I can do that. You want the duplicate attribute error to be emitted only if there's a single declared default, correct? Just want to make sure the precedence is done correctly.

Reply:
Ah, I see what I was missing before. It looks like right now we have this function, which checks that we don't have something like this:

```rust
#![feature(derive_default_enum)]
#[derive(Default)]
enum Foo {
    #[default]
    #[default]
    A,
    B,
}
```

and the multiple declared defaults error (*not* from this function) comes from:

```rust
#![feature(derive_default_enum)]
#[derive(Default)]
enum Foo {
    #[default]
    A,
    #[default]
    B,
}
```

I thought that these were actually referring to the same kind of bug (just checked in different places, but I see now that this isn't the case).

That said, it feels like the error message doesn't actually need to be different whether `#[default]` is put multiple times on the same variant or on multiple variants -- this error message basically "always works", IMO.

```
error: multiple `#[default]` attributes
 --> src/lib.rs:6:5
  |
4 |     #[default]
  |     ---------- `#[default]` used here
5 |     #[default]
  |     ---------- `#[default]` used again here
6 |     A,
  |     ^
  |
  = note: only one `#[default]` attribute is needed
help: try removing this
 --> src/lib.rs:5:5
  |
5 |     #[default]
  |     ^^^^^^^^^^
```

In any case, this now sounds potentially more complicated than what I was thinking, so happy to leave it for a future PR (just let me know). I do think there's an opportunity to merge these two checks into one (collect all the default attributes across all variants, if we end up with more than one, emit an error with appropriate spans), rather than splitting across two separate checks for on one variant and on multiple variants.

Reply:
I think there's value in clearly stating that only one variant can be default as opposed to just saying the attribute can only be used once. The other restriction is just because it's unnecessary, not that it would cause any semantic concern.

Reply:
I don't disagree, but it feels odd to have two separate sets of code to do this - we could still have a note as part of the error message about only one variant being the default or multiple default attributes on one variant being useless.

Reply:
Coming back around to this. You would prefer that the error message be merged, even though this would limit the ability to provide assists in the editor (namely an assist that remove the extraneous attributes and an assist setting the default). Is that correct? If so, I can do it, though I feel it necessary to state that I don't think the tradeoff is worth it.